streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -16,171 +16,256 @@
+            logPrefix
+        );
+
+        rebalanceInProgress = false;
     }
 
-    private void resumeSuspended(final Collection<TopicPartition> assignment) {
-        final Set<TaskId> suspendedTasks = partitionsToTaskSet(assignment);
-        suspendedTasks.removeAll(addedActiveTasks.keySet());
+    private void addNewTask(final Task task) {
+        final Task previous = tasks.put(task.id(), task);
+        if (previous != null) {
+            throw new IllegalStateException("Attempted to create a task that we already owned: " + task.id());
+        }
+    }
 
-        log.debug("Suspended tasks to be resumed: {}", suspendedTasks);
+    /**
+     * @throws IllegalStateException If store gets registered after initialized is already finished
+     * @throws StreamsException if the store's change log does not contain the partition
+     */
+    boolean updateNewAndRestoringTasks() {
+        final List<Task> restoringTasks = new LinkedList<>();
+        for (final Task task : tasks.values()) {
+            task.initializeIfNeeded();
+            if (task.state() == RESTORING) {
+                restoringTasks.add(task);
+            }
+        }
 
-        for (final TaskId taskId : suspendedTasks) {
-            final Set<TopicPartition> partitions = assignedActiveTasks.get(taskId);
-            try {
-                if (!active.maybeResumeSuspendedTask(taskId, partitions)) {
-                    // recreate if resuming the suspended task failed because the associated partitions changed
-                    addedActiveTasks.put(taskId, partitions);
+        boolean allRunning = true;
+        if (!restoringTasks.isEmpty()) {
+            changelogReader.restore();
+            final Set<TopicPartition> restored = changelogReader.completedChangelogs();
+            for (final Task task : restoringTasks) {
+                if (restored.containsAll(task.changelogPartitions())) {
+                    task.startRunning();
+                } else {
+                    // we found a restoring task that isn't done restoring, which is evidence that
+                    // not all tasks are running
+                    allRunning = false;
                 }
-            } catch (final StreamsException e) {
-                log.error("Failed to resume a suspended active task {} due to the following error:", taskId, e);
-                throw e;
             }
         }
+
+        // Note: if we wanted to process ready tasks while others are still restoring, all we have to do
+        // is delete this conditional and run the loop always.
+        if (allRunning) {
+            for (final Task task : tasks.values()) {
+                // TODO, can we make StandbyTasks partitions always empty (since they don't process any inputs)?
+                // If so, we can simplify this logic here, as the resume would be a no-op.
+                if (task.isActive()) {
+                    consumer.resume(task.inputPartitions());
```

Comment:
I cannot find the place where we pause partitions anymore.. we should pause all assigned partitions of the main consumer right after a rebalance when the thread is transition to Partitions_Assigned state.

Reply:
hmm. I thought we would now continue processing even during the rebalance, but just not try to commit? If that's true, we probably don't need to resume here anyway.

Reply:
If there are any restoring tasks, we want to pause on processing other "ready" tasks so that we can focus on restoring tasks -- of course if we move the changelog reader to another thread this is not an issue any more.